docs/architecture-decisions/2023-01-11_beaker_productionisation_review.md

```diff
@@ -0,0 +1,409 @@
+def increment(*, output: abi.Uint64):
+    """increment the counter"""
+    return Seq(
+        app.state.counter.set(app.state.counter + Int(1)),
+        output.set(app.state.counter),
+    )
+
+@app.external(authorize=Authorize.only(Global.creator_address()))
+def decrement(*, output: abi.Uint64):
+    """decrement the counter"""
+    return Seq(
+        app.state.counter.set(app.state.counter - Int(1)),
+        output.set(app.state.counter),
+    )
+
+```
+
+### (2) Defer PyTEAL compilation
+
+#### What?
+
+Currently, `beaker.Application.compile()` is called as part of `__init__()`, assuming there are no `precompiles` defined. We recommend that `compile()` always be deferred to a later point, and further that `compile()` does not mutate `Application` in any way, but instead returns a new object.
+
+#### Why?
+
+The deferment of the `compile()` call is actually a necessary part of recommendation #1 that we have skipped over thus far, but would be recommended anyway.
+
+The immediate `compile()` has issues such as requiring implementors (i.e. subclasses) to call `super().__init__()` as a **final** step in their own `__init__` method - any code that runs after the super init call will have no effect on the application produced!
+
+Immediate compilation also reduces the control the user has over the output. Although currently the only parameter that `compile` takes is a `client`, it might be useful to add (optional) parameters here to control the compilation. For example, if you can pass in the list of optimisations that should be applied, that allows you to have [output stability](https://github.com/algorandfoundation/algokit-cli/blob/main/docs/architecture-decisions/2022-11-22_beaker-testing-strategy.md#context) of your smart contract code if new optimisations are added in the future.
+
+The separation of compiled state outside of `Application` simplifies the design, and can be done mostly transparently to end-users.
+
+The separation of compiled state will also benefit future interoperability. It allows for more explicit decoupling of PyTEAL compilation (Beaker / PyTEAL transpilation -> TEAL) and deploment (TEAL -> compiled byte code -> Algorand network). Once `beaker.client` is split into a separate package, if the compiled state can be both generated from a beaker Application object _or_ loaded from disk (or similar), this means Beaker's ApplicationClient could be used in more situations, such as for a (say) tealish smart contract, or a C# smart contract, or a raw PyTEAL or TEAL smart contract, etc. This conforms better to the modularity principle in AlgoKit and also vice versa allows for a Beaker smart contract to be deployed by a TypeScript deployer, or C# deployer, etc.
+
+#### Before & After - user's perspective
+
+For most use cases, this should be a relatively small and probably imperceptible change.
+
+We believe there are two common usage scenarios that use the output of PyTEAL compilation currently:
```

Comment:
You may also be interested in metadata such as the mapping of source code to line numbers.

Reply:
Good point. I've added a note about that. I don't think it has an impact on the design though.